### PR TITLE
use current_schema in PLSQL::Schema#schema_name

### PR DESCRIPTION
**Pull request by substars**
We're using oracle-enhanced with schema: parameter set, so current_schema is different than the current user, which causes PL/SQL not to find available packages. current_schema should (?) be the same as session_user unless explicitly set otherwise, in which case we should use that.

I'm happy to write a specific test for this, but it'll make setup more complex since we'll need to create another user, do some grants, etc. Thoughts?

### CODE REVIEW

**Comment by jgebal** [comment]
I used alter session set current_schema a lot at my previous project as we were using multi-schema deployments on single dev instance to avoid stepping on each other's toes when developing. So each developer/feature had his schema to play with.
Seems like a good idea to vs able to mimic this in ruby-postal.

**Comment by jgebal** [comment]
From what I remember. We were using 
schema: schema_name in the database.html so all tests could run in a specified context per connection.

**Comment by jgebal** [comment]
This change looks good.
We could add setter method for schema name too.

**Comment by javornikolov** [comment]
Looks good to me. It may happen that in some contexts the *session user* would be needed, while in others - *current schema*.

> I'm happy to write a specific test for this, but it'll make setup more complex since we'll need to create another user, do some grants, etc. Thoughts?

It's good to have tests. I think we already have 2 test schema `hr` and `arunit` - would these be enough? It's not a huge problem to setup more users if needed.

**Comment by substars** [comment]
You're right, no reason not to write tests. Done!

File: spec/plsql/package_spec.rb [hunk 1]
```diff
@@ -20,7 +20,7 @@
         END test_procedure;
       END;
     SQL
-
+    plsql.execute("grant execute on TEST_PACKAGE to #{DATABASE_USERS_AND_PASSWORDS[1][0]}")
```
Comment (javornikolov) [inline spec/plsql/package_spec.rb]: That got me thinking about an edge case - what if we connect through [proxy user](http://www.pythian.com/blog/the-power-of-the-oracle-database-proxy-authenticated-connections/)? (I'm not sure whether the other tests would be compatible with such scenario either).

File: spec/plsql/package_spec.rb [hunk 2]
```diff
@@ -35,6 +35,24 @@
     expect(PLSQL::Package.find(plsql, :test_package)).not_to be_nil
   end
 
+  context "with current_schema set" do
+    before(:all) do
+      @old_connection = plsql.connection
+      plsql.connection = get_connection(1)
+      plsql.execute "ALTER SESSION set current_schema=#{DATABASE_USERS_AND_PASSWORDS[0][0]}"
+    end
+    it "should find existing package with current_schema set" do
```
Comment (javornikolov) [inline spec/plsql/package_spec.rb]: The convention is to have blank lines between `it .. end` blocks.

File: spec/plsql/package_spec.rb (hunk 1 above)
Comment (javornikolov) [inline spec/plsql/package_spec.rb]: > That got me thinking about an edge case - what if we connect through proxy user? (I'm not sure whether the other tests would be compatible with such scenario either).

Looks like the other tests already expect the username used to connect to match session user: 
[schema_spec.rb#L37](https://github.com/rsim/ruby-plsql/blob/e406170cf98f3e37721b828bd83179db01e604d2/spec/plsql/schema_spec.rb#L37).

So I think it's OK to rely on this assumption for now. We may think of how to address proxy user connections some other time in future.

File: spec/plsql/package_spec.rb [hunk 3]
```diff
@@ -35,6 +35,24 @@
     expect(PLSQL::Package.find(plsql, :test_package)).not_to be_nil
   end
 
+  context "with current_schema set" do
```
Comment (javornikolov) [inline spec/plsql/package_spec.rb]: One may say that `current_schema` is always set (by default to the connected user). So it might be good to be a bit more specific in the description here.

File: spec/plsql/package_spec.rb (hunk 2 above)
Comment (javornikolov) [inline spec/plsql/package_spec.rb]: I believe it's good to also have the scenario when the current_schema is not changed and it fails to find the package.

**Comment by substars** [comment]
Thanks for your feedback @javornikolo, I updated tests to be more descriptive (hopefully!).

Regarding proxy users and a setter for schema_name: the specs frequently reset connections just for the side effect of clearing cached metadata, which was confusing to me. oracle-enhanced sets current_schema right away on new connections so it is a relatively simple use case, but edge cases like those might also be easier to implement if the PLSQL::Schema API was more transparent about what it is caching (and how and when to reset it).

File: spec/plsql/package_spec.rb [hunk 4]
```diff
@@ -67,6 +66,43 @@
...
+        expect(PLSQL::Package.find(plsql, :test_package)).not_to be_nil
+      end
+
+      it "should report an existing procedure as existing" do
+        expect(plsql.test_package.procedure_defined?(:test_procedure)).to be_truthy
+      end
+
+    end
+
+    after(:all) do
```
Comment (javornikolov) [inline spec/plsql/package_spec.rb]: I wonder if would be better to have `after(:all)` close to the relevant `before` block (it feels a bit easier to see what we need to tear down).

**Comment by javornikolov** [comment]
Good spot about the caching... It may be cleaner to add a schema_name setter which takes care about purging the cached stuff. And still - looks like we're still in trouble when current_schema is changed by a database call, resetting the cache via `connection=` seems quite odd. I think we should add a public method for purging the cache; and maybe ability to turn on/off the caching.

**Comment by substars** [comment]
I agree that that's the way to go, but I think it's a separate issue and deserves its own PR. This is a surgical fix to get the gem working like I'd expect with oracle-enhanced 1.6, messing with caching is more tricky.

Re: the after block, you're right that this project usually has `after` adjacent to `before` so I changed mine to match. :octocat:

**Comment by javornikolov** [comment]
Cool, it looks good enough to me now :-) I agree - the other intricacies can be handled in separate PR.
So I'm merging this PR :shipit:

@substars, thank you very much for this contribution!
